packages/commerce-sdk-react/src/hooks/ShopperCustomers/cache.ts

```diff
@@ -4,94 +4,243 @@
-            update: [{queryKey: getCustomerPaymentInstrument.queryKey(newParams)}]
+            update: [
+                {
+                    queryKey: getCustomerPaymentInstrument.queryKey(newParams)
+                },
+                {
+                    queryKey: getCustomer.queryKey(newParams),
+                    updater: createUpdateFunction((customer: Customer) => {
+                        customer.paymentInstruments = [
+                            ...(customer.paymentInstruments ?? []),
+                            response
+                        ]
+
+                        return customer
+                    })
+                }
+            ]
         }
     },
     createCustomerProductList(customerId, {parameters}, response) {
         // We always invalidate, because even without an ID we assume that something has changed
-        // TODO: Rather than invalidate, can we selectively update?
-        const invalidate: CacheUpdate['invalidate'] = [
-            {queryKey: getCustomerProductLists.queryKey(parameters)}
-        ]
-        // We can only update cache for this product list if we have the ID
         const listId = response.id
-        if (!listId) return {invalidate}
         return {
-            invalidate,
-            update: [{queryKey: getCustomerProductList.queryKey({...parameters, listId})}]
+            update: [
+                {
+                    queryKey: getCustomerProductLists.queryKey(parameters),
+                    updater: createUpdateFunction((result: CustomerProductListResult) => {
+                        // Add new list to front of the lists.
+                        result.data.unshift(response)
+                        result.limit++
+                        result.total++
+
+                        return result
+                    })
+                },
+                {
+                    queryKey: getCustomerProductList.queryKey({...parameters, listId})
+                }
+            ]
         }
     },
     createCustomerProductListItem(customerId, {parameters}, response) {
         // We always invalidate, because even without an ID we assume that something has changed
-        // TODO: Rather than invalidate, can we selectively update?
-        const invalidate: CacheUpdate['invalidate'] = [
-            {queryKey: getCustomerProductList.queryKey(parameters)},
-            {queryKey: getCustomerProductLists.queryKey(parameters)}
-        ]
-        // We can only update cache for this product list item if we have the ID
+        // QUESTION: Why would we get a response that doesn't have an ID?
         const itemId = response.id
-        if (!itemId) return {invalidate}
+
         return {
-            invalidate,
-            update: [{queryKey: getCustomerProductListItem.queryKey({...parameters, itemId})}]
+            update: [
+                {
+                    queryKey: getCustomerProductListItem.queryKey({...parameters, itemId})
```

Comment:
According to the RAML spec, we could get a `response` with no `id`, meaning this update could potentially be storing data in the cache that we never actually retrieve (`itemId: undefined`).

In practice, though, I don't know when we'd ever not receive an `id`.